zokrates_core/src/flatten/mod.rs

```diff
@@ -2147,6 +2202,36 @@ impl<'ast, T: Field> Flattener<'ast, T> {
         }
     }
 
+    fn flatten_assertion(
+        &mut self,
+        statements_flattened: &mut FlatStatements<T>,
+        lhs: FlatExpression<T>,
+        rhs: FlatExpression<T>,
+    ) {
+        let lhs_flattened = match lhs {
+            FlatExpression::Number(_) | FlatExpression::Identifier(_) => lhs,
+            _ => {
+                let sym = self.use_sym();
+                statements_flattened.push(FlatStatement::Definition(sym, lhs));
+                FlatExpression::Identifier(sym)
+            }
+        };
+
+        let rhs_flattened = match rhs {
+            FlatExpression::Number(_) | FlatExpression::Identifier(_) => rhs,
+            _ => {
+                let sym = self.use_sym();
+                statements_flattened.push(FlatStatement::Definition(sym, rhs));
+                FlatExpression::Identifier(sym)
+            }
+        };
+
+        statements_flattened.push(FlatStatement::Condition(
+            FlatExpression::Mult(box lhs_flattened, box FlatExpression::Number(T::from(1))),
```

Comment:
I suspect we can be a bit more efficient here:

```
flatten_assertion(x * y, z)

currently:
t = x * z
t * 1 == z

optimal:
x * y = z
```

Some with the mirrored `flatten_assertion(z, x * y)` -> `x * y == z`

Something like

```
let (lhs, rhs) = match (lhs, rhs) {
   (FlatExpression::Mult(x, y), z) if x.is_linear() && y.is_linear() && z.is_linear() => (FlatExpression::Mult(x, y), z),
   (z, FlatExpression::Mult(x, y)) if x.is_linear() && y.is_linear() && z.is_linear() => (FlatExpression::Mult(x, y), z),
   (x, z) if x.is_linear() && z.is_linear() => (FlatExpression::Mult(x, 1), z)
   (x, z) if x.is_linear() => {introduce variable for z, (FlatExpression::Mult(x, 1), t)},
   (x, z) if z.is_linear() => {introduce variable for x, (FlatExpression::Mult(z, 1), t)},
   (x, z) => {introduce variables for both, FlatExpression::Mult(t, 1), u)}
}

```